plugins/modem-manager/fu-mm-device.c

```diff
@@ -0,0 +1,668 @@
+	 *     update procedure at this time, so we just exit without any
+	 *     detach.
+	 */
+
+        /* FuMmDevice created from MM... */
+	if ((self->omodem != NULL)) {
+                /* both fastboot and qmi-pdc supported? another write required */
+                if ((self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_FASTBOOT) &&
+                    (self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_QMI_PDC)) {
+                        g_debug ("both fastboot and qmi-pdc supported, so the upgrade requires another write");
+                        fu_device_add_flag (device, FWUPD_DEVICE_FLAG_ANOTHER_WRITE_REQUIRED);
+                }
+                /* fastboot */
+                if (self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_FASTBOOT)
+                        return fu_mm_device_detach_fastboot (device, error);
+                /* otherwise, assume we don't need any detach */
+                return TRUE;
+        }
+
+        /* FuMmDevice created from udev...
+         * assume we don't need any detach */
+        return TRUE;
+}
+
+typedef struct {
+	gchar  *filename;
+	GBytes *bytes;
+} FuMmFileInfo;
+
+static void
+fu_mm_file_info_free (FuMmFileInfo *file_info)
+{
+	if (file_info != NULL) {
+		g_free (file_info->filename);
+		g_bytes_unref (file_info->bytes);
+		g_free (file_info);
+	}
+}
+
+typedef struct {
+	FuMmDevice	*device;
+	GError		*error;
+	GPtrArray	*file_infos;
+	gsize		 total_written;
+	gsize		 total_bytes;
+} FuMmArchiveIterateCtx;
+
+static void
+fu_mm_qmi_pdc_archive_iterate_mcfg (FuArchive *archive, const gchar *filename, GBytes *bytes, gpointer user_data)
+{
+	FuMmArchiveIterateCtx *ctx = user_data;
+	FuMmFileInfo *file_info;
+
+	/* filenames should be named as 'mcfg.*.mbn', e.g.: mcfg.A2.018.mbn */
+	if (!g_str_has_prefix (filename, "mcfg.") || !g_str_has_suffix (filename, ".mbn"))
+		return;
+
+	file_info = g_new0 (FuMmFileInfo, 1);
+	file_info->filename = g_strdup (filename);
+	file_info->bytes = g_bytes_ref (bytes);
+	g_ptr_array_add (ctx->file_infos, file_info);
+	ctx->total_bytes += g_bytes_get_size (file_info->bytes);
+}
+
+static gboolean
+fu_mm_device_qmi_open (FuMmDevice *self, GError **error)
+{
+	self->qmi_pdc_updater = fu_qmi_pdc_updater_new (self->port_qmi);
+	return fu_qmi_pdc_updater_open (self->qmi_pdc_updater, error);
+}
+
+static gboolean
+fu_mm_device_qmi_close (FuMmDevice *self, GError **error)
+{
+	g_autoptr(FuQmiPdcUpdater) updater = NULL;
+
+	updater = g_steal_pointer (&self->qmi_pdc_updater);
+	return fu_qmi_pdc_updater_close (updater, error);
+}
+
+static gboolean
+fu_mm_device_write_firmware_qmi_pdc (FuDevice *device, GBytes *fw, GError **error)
+{
+	g_autoptr(FuArchive) archive = NULL;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(GPtrArray) file_infos = g_ptr_array_new_with_free_func ((GDestroyNotify)fu_mm_file_info_free);
+	FuMmArchiveIterateCtx iterate_context = {
+		.device = FU_MM_DEVICE (device),
+		.file_infos = file_infos,
```

Comment:
don't we need to zero set the other fields?

Reply:
hummm.... aren't all non-explicitly initialized members of a struct by default initialized to zero?

Reply:
Done